FEATURE: Replace existing assets with new version

radmiraal opened the pull request:
Introduce replacement of files in Neos. When an asset
is replaced the content cache for nodes that use the
asset or a variant is flushed. Also a 301 redirect from
the old resource url to the new url is created.

Because the changes are applied to the asset model directly
the changes will be live immediately. For this reason
the replace functionality is restricted to live publishers
in Neos.

In case image variants have cropping instructions they
will be reset to the maximum bounds for the new image.

robertlemke commented:
Some idea regarding the overview:

Is it easily possible to determine if assets are currently used and include that information into the media browser overview already? If an asset is used, we could deactivate the Delete button and create a little hover which says: "Can't delete this asset because it is in use.".

We could also add a new menu item to the new context menu" (i) Show usages" if it is currently used anywhere. That would show a popup with a list of document nodes where the asset is used.

robertlemke commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Strategy/AbstractAssetUsageStrategy.php:
```diff
@@ -0,0 +1,60 @@
+use TYPO3\TYPO3CR\Domain\Factory\NodeFactory;
+use TYPO3\TYPO3CR\Domain\Repository\NodeDataRepository;
+use TYPO3\Neos\Domain\Service\UserService as DomainUserService;
+
```
Lots of unused imports here
![image](https://cloud.githubusercontent.com/assets/95582/14703998/f2cb22aa-07b2-11e6-815d-4bd143b90712.png)

radmiraal replied:
jip, tried to clean them up while refactoring but need to rebuild my phpstorm index I think ;)

regarding the display of usages in the list view: this was also suggest in the change that implemented the usage view. I think it was @aertmann that raised a performance concern, and I think he's right there. Especially now we allow 3rd party packages to define if a resource is in use yes / no.

What we can do with the foldout menu is appending the delete button async, so we do a call to the server at the moment the menu is opened. Wdyt?

robertlemke replied:
Did I mention we need Event Sourcing and Projections? ;-)

Yes, I think it would be viable to do that async - but let's try it, because the user will need that information and shouldn't have to click again.

robertlemke commented on TYPO3.Neos/Classes/TYPO3/Neos/Controller/Module/Management/AssetController.php:
```diff
@@ -155,26 +157,18 @@ public function deleteAction(\TYPO3\Media\Domain\Model\Asset $asset)
+
+        /** @var AssetUsageInNodeProperties $usage */
+        foreach ($usageReferences as $usage) {
+            $relatedNodes[$usage->getSite()->getNodeName()]['site'] = $usage->getSite();
```
This results in a _Call to undefined method TYPO3\Neos\Domain\Strategy\AssetUsageInNodePropertiesStrategy::getSite()_ when I click on the _usages_ button in an asset's single view

radmiraal replied:
will fix

robertlemke commented:
I think the "replace asset" screen needs a bit more love, especially some text which explains what is about to happen.

![image](https://cloud.githubusercontent.com/assets/95582/14704125/74abe606-07b3-11e6-8a74-9e1f8d496c54.png)

How about starting with a headline "Replace asset" and then a little paragraph like

_"You can replace this asset by uploading a new file. Currently the asset is not in used anywhere on the website / Currently the asset is used in the following parts of your website:"_

Followed by a list of clickable document node links (open in a new tab?). If there are more then 10 documents, the list should probably be expandable ("show more" or "show all 231 usages").

The button for uploading a new file could be labeled "Choose new file"

radmiraal commented:
Agree to the points about the replace asset screen. will process them.

Regarding the list of references: would it already make sense to add checkboxes (checked by default) so the user can unselect references he doesn't want to touch?

And when we're rendering a list of references here this would (at the moment) be a list of usages in nodes. That's pretty much familiar stuff for us to handle, but when a 3rd party package would define that an asset is still in use... It is for example imported into a domain model. How to handle that case?

I now see 2 options, one being less work... the other one needing more thought for sure:
1) We can simply display the count of references that will also be touched, and not allow to choose
2) We can try to search for a generic way to render a table that displays the usage and allow the user to choose.

I'm now opting for doing the first option because showing different kinds of references will give a lot of challenges when it comes to listing a table :-)

Thinking about the first option I realize we might need strategies for replacement of the resource too

robertlemke commented:
I'd also go for option 1) first. If assets are used somewhere else than in nodes, I think it's okay to display a generic message for the time being.

radmiraal commented:
Updated the change so the 'delete' button is disabled in the list view if the asset is in use. This is the same behavior as got merged into master a few weeks ago: as long as the asset is in use it can not be removed. To add the details about the usage to the foldout menu I added an extra option 'x usages' to the foldout menu if it is in use.

Try to ignore the styling please, but it looks kind of like:
![screenshot 2016-04-26 19 15 31](https://cloud.githubusercontent.com/assets/324613/14827884/c84aa36c-0be4-11e6-856b-4b5cc6f812ba.png)
![screenshot 2016-04-26 19 15 21](https://cloud.githubusercontent.com/assets/324613/14827890/cffca010-0be4-11e6-8c76-8f9b450fcdce.png)

1) What do you think about this 'flow'?

2) In the list view you see the foldout menu on the ellipsis horizontal icon (3 dots). I added that closer to the name of the file (similar to the UI of owncloud) as those actions are probably more used than the metadata showed next to it. I didn't remove the existing icons (edit and remove) because that might confuse users. I'm just not yet sure if:
- removing them would be wiser
- the ellipsis icon should be the column next to the filename or moved to the right (before the edit icon)

radmiraal commented:
Substracted some of the work in smaller PR's already: https://github.com/neos/neos-development-collection/pull/483 https://github.com/neos/neos-development-collection/pull/482 https://github.com/neos/neos-development-collection/pull/481

Also moved the strategy change to a separate branch (depends on https://github.com/neos/neos-development-collection/pull/483 ): https://github.com/radmiraal/neos-development-collection/commit/304e924908b51c4427968fac295292e604f65ac2

Those changes can already be reviewed / made merge ready so we can make this PR just about the replacement feature

aertmann commented:
Hey looks nice, have a few points to add

we considered disabling the delete button on the list views when adding the overview, but I advised against it because the query to find the usages is really slow and will be a performance nightmare on sites with lots of nodes (see https://github.com/neos/neos-development-collection/blob/e580ad276e62fc5084d5f9b72e22ad082174b4b8/TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php#L1384)

Quite some effort has gone into making the media library work with lots of assets/nodes and that small addition doesn't justify making it much slower to render IMO. We've talked about making an actual reference table of the references, which would allow doing it without the performance impact. Until then we have to live with an error.

also AFAIK it's nothing new that asset cannot be deleted if there are in use, that has been in place since 1.1 (see https://github.com/neos/neos/blob/1.1/Classes/TYPO3/Neos/Controller/Module/Management/AssetController.php#L57)

regarding the icons, I'd suggest adding the dots at the end.. maybe remove the edit/remove icons if it feels better..

radmiraal commented:
Hey @aertmann: thanks for the feedback ;-)

Regarding the disabling of the delete button: I've seen your argument, and I agree we should keep it as light as possible. But what are the alternatives? We can try lazy loading which could be done for all displayed assets on load, but that would again be very specific. Doing a lazy load of the available actions per asset when opening the context menu would also be possible, but that would again introduce a delay.

With https://github.com/radmiraal/neos-development-collection/commit/304e924908b51c4427968fac295292e604f65ac2 I think the change is not that heavy. It only loads the usage once per asset, and only if the usage is used in the view. But still this could have a performance hit with a huge amount of nodes ofc. 

So basically the question is: will we call {asset.inUse} or {asset.usageCount} in the list view yes or no, and would there be a difference between the media selection from the inspector or the actual media view... @robertlemke you any thoughts on the topic now we have Askes concerns in the discussion?

robertlemke commented:
Question is: how much effort is it to create a reliable projection table just for looking up the references / usage count, like @aertmann mentioned? From UX point of view, it would be far better having the information in place for the user interface, also for other scenarios, for example if we want to implement multi-selection for deletion (choose multiple assets and delete them at once).

If a dedicated table for reference information purposes makes sense to you and can be reliably kept up to date, I'd say, let's try this (if it only takes a few hours).

robertlemke commented:
Further feedback regarding the user interface:

![image](https://cloud.githubusercontent.com/assets/95582/14951876/4ec05c9a-105c-11e6-974c-5a1520898ede.png)

- I'd put the `Delete` menu entry below the usages and wouldn't add `(in use)` to the label
- for the usages: how about "Show usages (6)" instead of "6 usages"? As for the icon, how about just using `info` or `info-circle`?
- needs fine tuning for positioning / font sizes etc for the new menu items

![image](https://cloud.githubusercontent.com/assets/95582/14952037/59b2aa44-105d-11e6-9965-aaf44f1f6646.png)

For the "Usages " screen:

- change headline to from `References to "KeyVisual.jpg" to `Usages of "KeyVisual.jpg"`
- show a thumbnail between the main headline and the table, so you get immediate feedback which asset it is about
- from the example in my screenshot it doesn't get clear what the "Reference" column is about, or?

![image](https://cloud.githubusercontent.com/assets/95582/14952109/dacab9a0-105d-11e6-8d39-88adc54a16f2.png)

For the "Replace" screen:

- should generally be consistent with other screens in terms of headline etc, so for example:
- add a headline (similar to the Usages screen), for example `Replace "KeyVisual.jpg"`
- just remove the "Upload a new file" headline, but keep the short intro text
- extend the text a bit: `You can globally replace this asset by uploading a new file. Once replaced, the new file will be used instead of the old one everywhere on your website.`
- maybe just drop the headline "Options" as long as we only have one?
- instead of `File usage` just `Usage`?
- replace the text `Currently the asset is used, click the link below to see details about the usage.` by `This asset is currently used 21 times.`
- change the button label to `Show all usages`?

radmiraal commented:
Hey Robert, made some changes, see screenshot. It shows the replace asset screen next to the edit user screen.

Regarding the options: I hope that we introduce multiple options soonish (like create redirects yes / no, and so on), but for now it's indeed better to remove the label.

![screenshot 2016-05-03 10 51 09](https://cloud.githubusercontent.com/assets/324613/14978592/2c53b5d2-111d-11e6-9ff7-afef363a74ad.png)

radmiraal commented:
added support for creating 301 redirects on replacement if the redirecthandler package is installed

robertlemke commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;ny5aN6bM&#x2F;1-replacement-of-assets-pictures"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Replacement of assets &#x2F; pictures</a>

robertlemke commented:
@radmiraal regarding the headlines - I'd just do it the other way around and make `Replace "alice-1.jpg"` bigger and `New file` and `Usage` smaller. What do you think?

![image](https://cloud.githubusercontent.com/assets/95582/15119811/fc0326b0-1612-11e6-81bb-44728d12502c.png)

robertlemke commented:
How much effort would it be to load the usages screen via Ajax and display it within the replace dialog screen instead of jumping to a new URL and screen? I think, from a user's perspective it would be less confusing if that could be done.

![image](https://cloud.githubusercontent.com/assets/95582/15119869/40451b9e-1613-11e6-93c2-ea6337d22f0d.png)

robertlemke commented:
One thing which is a bit confusing: when you go to the Media Browser through the `Media` button in the Inspector ("extended inspector view") and replace an asset by a new one which is still visible on the current page, the old asset will of course be visible after the replacement. This could be even more confusing if you use the Cropping Tool because it will show the old asset but applies the cropping to the new asset.

We should reload the current document (like in the Inspector with reloadOnChange) when an asset has been replaced.

robertlemke commented:
The edit dialog seems to be broken.

- go to the Media Browser
- click on "Edit" in the new context menu of an asset
- fatal exception:

```
Router resolve(): Could not resolve a route for building an URI for the given route values.
    asset => object: TYPO3\Media\Domain\Model\Image
    @action => replaceassetresource
    @controller => asset
    @package => typo3.neos
```

URL of the request causing the error is: ```http://neos-master.dev:8080/neos/content/images/edit.html?asset%5B__identity%5D=a0178d77-5c04-b829-a613-c9d2b6fb92cb```

radmiraal commented:
Fixed 2 bugs and updated the screen to look like:
![screenshot 2016-05-12 23 23 59](https://cloud.githubusercontent.com/assets/324613/15230958/b1f1c058-1898-11e6-8ef4-9fb483a59fa1.png)

Working on the refresh / ajax stuff, but would suggest doing that in a separate PR. @robertlemke: If you agree to the last 3 commits I can squash them and remove the WIP from the change ;-)

radmiraal commented:
went through the code and did some minor doc comment tweaks and translated a flash message. Now squashed the commits and removed WIP. Will still continue on some improvements in separate changes (like the reloading of the preview image in the inspector which turns out to be more complex than hoped)

robertlemke commented:
Hey thanks! While we're at it, let's quickly add a `Cancel` button in this screen, ideally with the ability to close the dialog by pressing ESC:

![image](https://cloud.githubusercontent.com/assets/95582/15316544/cbc61f3c-1c1d-11e6-900b-6cea7d14accf.png)

robertlemke commented:
Allright, so when I replace the wallpaper image of the demo site, of course the preview in the media browser isn't updated as we discussed earlier. When I reload the page, I do see the new image as the wallpaper, but the processed image and also the thumbnail has the wrong size - supposedly using the dimensions of the old image for calculating the thumbnail?

![image](https://cloud.githubusercontent.com/assets/95582/15317051/1aa0c1dc-1c20-11e6-8e03-714815fe6a2a.png)

When I run `media:clearthumbnails` and reload the page, the inspector thumbnail is still there, but the wallpaper in the content area is gone:

![image](https://cloud.githubusercontent.com/assets/95582/15317104/52d0e168-1c20-11e6-9751-8637acb8e5fe.png)

When I open the page in a new window, the wallpaper _is_ there, in the dimensions like in the first screenshot.

When I go to the crop editor in the inspector and move the box a little, then press apply, all caches are updated correctly and I see the correct wallpaper:

![image](https://cloud.githubusercontent.com/assets/95582/15317189/b8a455ec-1c20-11e6-8c07-23dc0effbf58.png)

robertlemke commented:
Regarding the reloading of the preview in the inspector: As hard as it is, but I think it needs to be part of this pull request, because if we would merge this PR without the fix, we'd introduce a rather confusing behaviour which could, as per our definition of a stable master, be released any time without the following fix. I think we need to play safe and include the fix already - what do you think?

aertmann commented on TYPO3.Neos/Resources/Public/JavaScript/Content/Inspector/Editors/ImageEditor.js:
```diff
@@ -298,9 +298,18 @@ function (Ember, $, FileUpload, template, cropTemplate, BooleanEditor, Spinner,
+
+			try {
+				value = JSON.parse(this.get('value'));
+			} catch (exception) {}
```
is this really necessary? why would it ever be invalid json? isn't there another way to check if it should be parsed or not?

radmiraal replied:
It would only fail if the value is empty. But why wouldn't we do this defensive style? As if, for some unforeseen reason, the value wouldn't be json (maybe a faulty custom editor or whatever) we would have an uncaught exception

aertmann replied:
hmm in that case I'd actually prefer an exception instead of it silently being swallowed.. I'd go with s.th. like instead, but guess that's a matter of preference.. anyway I'd imagine a lot of other stuff breaking before with an invalid value anyway

    value = this.get('value') ? JSON.parse(this.get('value')) : null;

robertlemke replied:
@aertmann how about logging that to the JS console but still ignoring it? Actually, the user can't really do anything to solve the problem and might be better off without an exception.

aertmann commented on TYPO3.Neos/Resources/Public/JavaScript/Content/Inspector/Editors/ImageEditor.js:
```diff
@@ -311,6 +320,27 @@ function (Ember, $, FileUpload, template, cropTemplate, BooleanEditor, Spinner,
+				assetChosen: function(assetIdentifier) {
+					if (assetIdentifier) {
+						this._assetIdentifier = assetIdentifier;
+						this._reloadPreviewImage(this._assetIdentifier);
```
this still causes the image to be reloaded even if it's already chosen right?

radmiraal replied:
yes, as it was before

aertmann commented on TYPO3.Neos/Resources/Public/JavaScript/Content/Inspector/Editors/ImageEditor.js:
```diff
@@ -311,6 +320,27 @@ function (Ember, $, FileUpload, template, cropTemplate, BooleanEditor, Spinner,
+					}
+				},
+				onLoad: function(event) {
+					if (this._assetIdentifier) {
```
reading the code, this would always reload the image when the iframe has loaded right if there's an image already selected that is, or?

radmiraal replied:
yes, couldn't find a way around that. It's too many reloads of the image, but I can not find a way to just listen to POST requests.

aertmann replied:
hmm I'd say we need to, that's definitely a step back compared to before to fix something that happens very rarely where the new unnecessary reload would happen all the time

aertmann commented on TYPO3.Neos/Resources/Public/JavaScript/Content/Inspector/Editors/ImageEditor.js:
```diff
@@ -298,9 +298,18 @@ function (Ember, $, FileUpload, template, cropTemplate, BooleanEditor, Spinner,
 			window.Typo3MediaBrowserCallbacks = {
-				assetChosen: function (assetIdentifier) {
+				_assetIdentifier: value && '__identity' in value ? value.__identity : null,
+				_reloadPreviewImage: function(assetIdentifier) {
```
the value doesn't need to be parsed does it? it's available as `this._assetIdentifier` right?

robertlemke commented:
This checkbox here looks a bit funny:

![image](https://cloud.githubusercontent.com/assets/95582/15357216/c0e9d6d6-1cfd-11e6-8886-d9ab12ce0a6a.png)

the one above seems to be the unchecked version and this one the checked one:

![image](https://cloud.githubusercontent.com/assets/95582/15357227/d8d408ca-1cfd-11e6-955c-3a43b89db7fc.png)

Me thinks that the unchecked version should have no check mark.

kdambekalns commented:
@robertlemke that is the styling since 2.2 (see https://github.com/neos/neos-development-collection/pull/457), not related to this PR.

radmiraal commented:
the checkmark thing: that's how the Neos UI is, it's a default component... should be a different issue

radmiraal commented:
I changed the list with references so they don't link by default, but allow to open the reference in an external window. I don't see a quick and clean way how we could make this work from the media browser and media module, and the current way completely disrupts the flow.

@see: 
![screenshot 2016-05-18 14 42 23](https://cloud.githubusercontent.com/assets/324613/15359073/e485c816-1d07-11e6-8e47-44e0bebceb11.png)

robertlemke commented:
Great improvements! So, just a little bugger, not sure if we can solve that easily: when you replace an asset which is currently also shown in the inspector, you correctly reload the inspector and I can see the picture I just replaced. But that also (more or less correctly) activates the inspector's Apply button, even though, server-side-technically there is no change (browser reload solves it). So, is it possible or too hacky to reset the dirty state of the inspector so that the Apply button doesn't become active?

radmiraal commented:
LOL :-) I actually made sure the apply button becomes active because otherwise the page reload is not triggered after closing the media browser, which would leave the old image in the rendered page.
Now the editor has to click 'apply' and the reloadAfterChange is triggered causing the page content to be reloaded with the new image

robertlemke commented:
Allright, I think that's a trick we can do. But let's try to only active the Apply button when something actually changed. For example, when I choose edit of an asset in the media browser, but then cancel, the Apply button also will become active.

robertlemke commented:
Another note regarding security: since the replacement of an asset is global, I think we should add two things:

1. restrict the replacement of assets to roles which have write access on the live workspace
2. make it clear in the replace dialog that the replacement is global, across all published or unpublished workspaces:

> You can globally replace this asset by uploading a new file. Once replaced, the new asset will be used on all places where the asset is used on your website.
> **Note**: This operation will replace the asset in all published or unpublished workspaces, including the live website.

robertlemke commented:
The new access restriction works fine, thanks!

In the references list, I wondered about the open-in-new-window link:

![image](https://cloud.githubusercontent.com/assets/95582/15362786/bf8368a2-1d16-11e6-8721-8f6a136e430a.png)

- why is Neos only showing one link?
- the link incorrectly points to the Latvian version, not the English version in John Doe's workspace

Maybe an issue in combination with multiple workspaces?

radmiraal commented:
ah, no... my bad, will fix... think I pulled it out of a foreach loop

robertlemke commented:
Thanks, the links work fine now! Sorry, for still nagging you, but I think the table shouldn't turn all blue on mouse over, or?

![image](https://cloud.githubusercontent.com/assets/95582/15364033/2798cfb4-1d1b-11e6-82ac-000734e92b41.png)

robertlemke commented:
Hey, that works!

@radmiraal if you really like to squash, that would be the time.
@aertmann I guess we can merge this now. I mean, not that I'm out of ideas how to improve this whole feature, but it's a good version 1!

radmiraal commented:
There's certainly room for improvement, but loads of those improvements are visible because of this change but furthermore unrelated. Ifthere's issues please put them in jira and ping me, as I've gone through the code a lot now it might be easy to fix them quickly...

Squashed the commits and updated the PR description.

radmiraal commented:
btw, @aertmann: the amount of thumbnail refreshes is already reduced to once every time the index is opened except the first time. Further reduction probably needs passing around action arguments or something similar which I don't think is the thing to do in this change

kdambekalns commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Service/AssetService.php:
```diff
@@ -218,4 +226,77 @@ public function validateRemoval(AssetInterface $asset)
+
+        if ($redirectHandlerEnabled) {
+            /** @var \Neos\RedirectHandler\Storage\RedirectStorageInterface $redirectStorage */
+            $redirectStorage = $this->objectManager->get('Neos\\RedirectHandler\\Storage\\RedirectStorageInterface');
```
Why not ``get(\Neos\RedirectHandler\Storage\RedirectStorageInterface::class)``? Works even if the class is not actually available, but does away with a string…

radmiraal replied:
wasn't that at some point posted as possibly unwanted behaviour of php? Because I only did it for the case the class wouldn't be there, so can change it

kdambekalns commented:
Didn't try but gave the code a quick look. Awesome feature, thanks @radmiraal! ;)

robertlemke commented:
I'm ready to merge this. Anyone not feeling like it should be merged? @aertmann

aertmann commented:
@robertlemke: a bit trigger hasty with the merge button here, especially considering the size of the PR.. haven't done a full review yet, but was planning to do that today and see if it there was anything I could improve.

Merging as first version is fine, as long as it is release ready. But not sure if you consider it that or not yet, I don't. Using a feature-branch would be better in this case since it could be merged, but only go into master when it's finalized. But not have the same restriction of only being a personal fork.

Anyway 👍 to @radmiraal for getting through all the comments, since it got sidetracked quite a lot with stuff that was already or unrelated to this PR.

Will comment this PR, then those need to be fixed in a new PR.

aertmann commented on TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php:
```diff
@@ -259,6 +268,42 @@ public function newAction()
+     * @param array $options
+     * @return void
+     */
+    public function updateResourceAction(AssetInterface $asset, FlowResource $resource, array $options = [])
```
I find it a little inconsistent this is named `updateResource` when the other one is named `replaceAssetResource`

aertmann commented on TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php:
```diff
@@ -259,6 +268,42 @@ public function newAction()
+     * @param Asset $asset
+     * @return void
+     */
+    public function replaceAssetResourceAction(Asset $asset)
```
is it on purpose it's not `AssetInterface`?

aertmann commented on TYPO3.Media/Resources/Private/Templates/Asset/ReplaceAssetResource.html:
```diff
@@ -0,0 +1,59 @@
+{namespace m=TYPO3\Media\ViewHelpers}
```
labels not translatable

aertmann replied:
labels not translatable

radmiraal replied:
just like 80% of the other labels in the file, probably best to create an issue to check all those templates

aertmann commented on TYPO3.Neos/Resources/Private/Partials/Media/ListView.html:
```diff
@@ -81,8 +80,32 @@
+									</f:security.ifAccess>
+									<li>
+										<a href="#" class="neos-media-delete" title="{deleteAssetLabel}" data-neos-toggle="tooltip" data-modal="delete-asset-modal" data-label="{asset.label -> f:format.crop(maxCharacters: 50)}" data-object-identifier="{asset -> f:format.identifier()}">
+											<i class="icon-trash icon-white"></i> Delete
```
labels not translatable

radmiraal replied:
see above

aertmann commented on TYPO3.Neos/Resources/Private/Partials/Media/ThumbnailView.html:
```diff
@@ -2,16 +2,46 @@
+									</f:security.ifAccess>
+									<li>
+										<a href="#" class="neos-media-delete" title="{deleteAssetLabel}" data-neos-toggle="tooltip" data-modal="delete-asset-modal" data-label="{asset.label -> f:format.crop(maxCharacters: 50)}" data-object-identifier="{asset -> f:format.identifier()}">
+											<i class="icon-trash icon-white"></i> Delete
```
labels not translatable

radmiraal replied:
see above

aertmann commented on TYPO3.Neos/Resources/Private/Templates/Media/Asset/ReplaceAssetResource.html:
```diff
@@ -0,0 +1,77 @@
+{namespace m=TYPO3\Media\ViewHelpers}
```
labels not translatable

aertmann replied:
labels not translatable

radmiraal replied:
see above

aertmann commented on TYPO3.Neos/Resources/Private/Styles/Media/_Media.scss:
```diff
@@ -643,6 +595,27 @@
+
+					.neos-caption {
+						padding: 0 0 0 .5rem;
+						margin-right: 2rem;
```
`rem` shouldn't be used in general, all sizes are defined in pixels and usually relative to `$defaultMargin`, `$relatedMargin`, `$tightMargin` or `$unit`

aertmann commented on TYPO3.Neos/Resources/Private/Partials/Media/ThumbnailView.html:
```diff
@@ -2,16 +2,46 @@
 {namespace neos=TYPO3\Neos\ViewHelpers}
```
labels not translatable

aertmann commented on TYPO3.Neos/Resources/Private/Partials/Media/ThumbnailView.html:
```diff
@@ -2,16 +2,46 @@
+							<div class="neos-dropdown-menu-list" role="menu">
+								<ul>
+									<li>
+										<f:link.action action="edit" arguments="{asset: asset}" title="{neos:backend.translate(id: 'media.tooltip.editAsset', source: 'Modules', package: 'TYPO3.Neos')}">
```
labels not translatable

radmiraal replied:
see mention about translation above

aertmann commented on TYPO3.Neos/Resources/Private/Styles/Foundation/_utilities.scss:
```diff
@@ -43,3 +43,7 @@
 }
+
+.neos-buffer-below {
+  margin-bottom: 1em;
```
$unit

aertmann commented on TYPO3.Neos/Resources/Private/Styles/Media/_Media.scss:
```diff
@@ -708,6 +681,11 @@
 
+	.neos-media-content-help {
+		font-style: italic;
+		padding-bottom: 1em;
```
`$defaultMargin`?

aertmann commented on TYPO3.Neos/Resources/Private/Translations/en/Modules.xlf:
```diff
@@ -209,6 +209,13 @@
 			</trans-unit>
 
+			<trans-unit id="media.dragHelp" xml:space="preserve">
+				<source>Drag and drop image on a collection / tag to add them to it.</source>
```
not just images, assets

radmiraal commented:
@aertmann Not going into the fact if you've valid points yes / no, but with the info in the PR (4 people looked at it including you, and there was no mention about your plan to do a longer session today) I suggest to skip a bit of the tone.

robertlemke commented:
Actually, I think @aertmann is right. A PR of this size deserves more time for reviewing the final version. We had some time pressure on our side and I think the PR could be released as is, but that shouldn't trigger my trigger-happyness.

Thanks, @radmiraal for developing this great feature and @aertmann for all the effort reviewing it!

radmiraal commented:
As said, I'm not saying something about being valid points. Just think exactly those kind of responses lead to not merging anything and lead to people like me leaving.

Regarding the code comments: have most of them fixed locally, just need to fix the sass variable stuff for margins. For the translations: that's another sidetrack tbh, the full templates are full of labels which are not yet translatable and I suggest to create an issue for that instead of just fixing the current mentions.

aertmann commented:
Left some inline comments, nice work @radmiraal, mostly just some nitpicks. But the missing label translation is a release blocker.

Btw. no documentation is needed for this, right?

![image](https://cloud.githubusercontent.com/assets/903567/15383820/5c1ad5ac-1dd9-11e6-967b-6d7d7990b326.png)
Looks a bit weird with more padding over than below..

![image](https://cloud.githubusercontent.com/assets/903567/15383831/7307c2ca-1dd9-11e6-9d95-1fbbf9d8f6d6.png)
The file type icon looks really bad on a retina display. Should be `32px` instead and then sized to `16px`

![image](https://cloud.githubusercontent.com/assets/903567/15383937/2a653222-1dda-11e6-8e3a-a83bd00d97bf.png)
Why is the font-size smaller?

![image](https://cloud.githubusercontent.com/assets/903567/15383984/7ef2f9f0-1dda-11e6-8169-ffa084e3e796.png)
The dropdown menu should also be present in the list view right?

![image](https://cloud.githubusercontent.com/assets/903567/15384165/adeace1c-1ddb-11e6-8fad-af693605a9ff.png)
I kinda overlooked this completely, maybe we should make that more clear. Especially the information about how many times it's used.

![image](https://cloud.githubusercontent.com/assets/903567/15384214/1c9dd570-1ddc-11e6-872f-74e8dc82f89e.png)
I get this because the Crowdin stuff is not set up yet, so there is only a translation file for `en`.

When you delete an asset that is in use through the media browser, you don't get the error notification
(existing new issue)

![image](https://cloud.githubusercontent.com/assets/903567/15384300/e7d943e6-1ddc-11e6-986a-8a68e8ee21a2.png)
When clicking the "Replace" button when no file has been selected.

![image](https://cloud.githubusercontent.com/assets/903567/15384333/16ef2f7e-1ddd-11e6-84ae-10d0b966d008.png)
Get this when showing the usages of an asset through the media browser.

![image](https://cloud.githubusercontent.com/assets/903567/15384594/ff92f732-1dde-11e6-8f4b-32966bb1962c.png)
Get this when I click the image twice of an image that's already selected (it should close the media browser).

![image](https://cloud.githubusercontent.com/assets/903567/15384619/42feeb2a-1ddf-11e6-98b0-600a3b572e15.png)
Menu is cut-off on the last asset in a row.

![image](https://cloud.githubusercontent.com/assets/903567/15384758/3bcddc0c-1de0-11e6-9d99-a4ca174ff2fe.png)
When clicking the thumbnail of an image, the cancel button is now present. If clicked you get back to the overview and and clicking an image then has this error.

Similarly, when clicking the thumbnail and then "Replace", clicking cancel goes to the overview instead of the image details. Solution would be to use the `history.back(-1)` on the replace cancel button like in the usages view.

![image](https://cloud.githubusercontent.com/assets/903567/15384828/bbe35d86-1de0-11e6-8cfd-5960d13ea3f5.png)
Would like to try if this could be tuned a little to look more like the rest of the UI.

aertmann commented:
@radmiraal: Sorry for the tone, it was my way of saying I wished I would have had a chance to voice my opinion before it was merged. The fact that I had planned that was irrelevant, but an example of why it's good to wait a bit. Christian often does this where he gives it 24 hours to let people object and Robert even did, but due to timezone differences I couldn't respond quickly enough. This especially applies to large PRs like these, which btw. I would like to avoid in the future by using feature-branches that we can collaborate on together and merge into master when we've sorted out all known issues or justified them going into master.

aertmann commented:
@radmiraal: as for the missing translations, didn't know it was that bad, so ignore that part then

radmiraal commented:
@aertmann curious: should issues that were present in master before this PR be created as jira tickets? As I think most of them were already present like the "focusable" error that clearly happens in the 'new file' screen too as required="required" triggers that error in Chrome?

aertmann commented:
@radmiraal: hmm good point, didn't that one already existed.. but I'd say as you prefer, just pointed out things and we can then decide to fix it them related to this or create new tickets for them.. I could spend some time tomorrow looking into some of them, while they're fresh in memory

new issues that didn't exist would of course be preferable to fix asap, but again it's a bit about how much effort it takes to create an issue compared to fixing it right away

aertmann commented:
@radmiraal: btw. regarding the labels, I remembered it as most of them being translated so checked again and it does seem like they are.. except the labels in PHP you fixed recently.. the thing is that it's a little tricky because they are only translated in the overwritten templates in Neos, so all the templates in Media are not translated.. so we might have to copy the templates to Neos and fix it there :(

most of it was fixed in https://github.com/neos/neos-development-collection/pull/387, but we decided to skip the Media templates since they are going away soon anyway

radmiraal commented:
@aertmann:

regarding the required="required" => that happens as Chrome kan not do HTML5 validation on hidden fields, which is kind of an issue with the file uploader.

regarding the labels: I was thinking about using a partial for the translations, that way we could overwrite just the partial (that would call the translate viewhelper) in the Neos package, and get back to a single template for both packages... Besides that I really want to separate the mediabrowser controller from the media module controller...

I'll push the initial set on feedback to a PR as they make sense to get into master soonish, what about the other issues here? Maybe open a feature branch for media management improvement and hack the world away the next days? I would spend some time on the topic anyway as I've gone so deep I shouldn't step out too soon ;-)

kdambekalns commented:
Thanks @radmiraal for pushing this further!

As far as the separation goes, when @aertmann wrote "soon" I had to smile. Been there, done that… Soon is always later, so why not fix it? At least if someone would do it "voluntarily"… 👼 

And 💯 to the idea of squashing those little beasts in a branch! Keep it going! ❤️

aertmann commented:
@radmiraal: 👍 for a media related branch

regarding labels, that is a good idea but won't solve the issue since there's a lot of differences in the html as well.. basically I'd suggest to ignore the default templates (they are broken anyway) and just focus on the Neos templates and then that becomes the templates of the new media browser package with all the related labels, controllers and such..

@kdambekalns: wasn't saying not to fix something, just to ignore the media templates since they are broken, not used and going away anyway

also #159 already exists and since @radmiraal and @dfeyer both would like to see this, we should try to achieve that before `2.3`.. although I would wait until we have fixed the stuff related to this change and then do it..

aertmann commented:
created a new branch and WIP PR for it, see https://github.com/neos/neos-development-collection/pull/532 / https://github.com/neos/neos-development-collection/tree/media-browser
